apps/passport-server/src/services/telegramService.ts

```diff
@@ -102,110 +103,130 @@ export class TelegramService {
-              userId,
-              `<i>You left ${chat?.title}. To join again, you must re-verify by typing /start.</i>`,
-              { parse_mode: "HTML" }
+      return traced("telegram", "chat_member", async (span) => {
+        try {
+          const newMember = ctx.update.chat_member.new_chat_member;
+          span?.setAttribute("userId", newMember.user.id.toString());
+          span?.setAttribute("status", newMember.status);
+
+          if (newMember.status === "left" || newMember.status === "kicked") {
+            logger(
+              `[TELEGRAM] Deleting verification for user leaving ${
+                newMember.user.username || newMember.user.first_name
+              } in chat ${ctx.chat.id}`
+            );
+            await deleteTelegramVerification(
+              this.context.dbPool,
+              newMember.user.id,
+              ctx.chat.id
             );
+            const chat = await getGroupChat(ctx.api, ctx.chat.id);
+            span?.setAttribute("chatId", chat.id);
+            span?.setAttribute("chatTitle", chat.title);
+
+            const userId = newMember.user.id;
+            if (!newMember.user.is_bot) {
+              await this.authBot.api.sendMessage(
+                userId,
+                `<i>You left ${chat?.title}. To join again, you must re-verify by typing /start.</i>`,
+                { parse_mode: "HTML" }
+              );
+            }
           }
+        } catch (e) {
+          logger("[TELEGRAM] chat_member error", e);
+          this.rollbarService?.reportError(e);
         }
-      } catch (e) {
-        logger("[TELEGRAM] chat_member error", e);
-        this.rollbarService?.reportError(e);
-      }
+      });
     });
 
     // The "start" command initiates the process of invitation and approval.
     this.authBot.command("start", async (ctx) => {
-      const userId = ctx?.from?.id;
-      try {
-        // Only process the command if it comes as a private message.
-        if (isDirectMessage(ctx) && userId) {
-          const username = ctx?.from?.username;
-          const firstName = ctx?.from?.first_name;
-          const name = firstName || username;
-          await ctx.reply(
-            `Welcome ${name}! 👋\n\nClick the group you want to join.\n\nYou will sign in to Zupass, then ZK prove you have a ticket for one of the group's events.\n\nSee you soon 😽`,
-            { reply_markup: zupassMenu }
-          );
+      return traced("telegram", "start", async (span) => {
+        const userId = ctx?.from?.id;
+        if (userId) span?.setAttribute("userId", userId?.toString());
+        try {
+          // Only process the command if it comes as a private message.
+          if (isDirectMessage(ctx) && userId) {
+            const username = ctx?.from?.username;
+            if (username) span?.setAttribute("username", username);
+            const firstName = ctx?.from?.first_name;
+            const name = firstName || username;
+            await ctx.reply(
```

Comment:
ctx reply may be another thing worth wrapping in a function that takes a ctx and a message, and executes ctx.reply in a `traced`

Reply:
Agreee